syncop: use flexible array for synctask's stack (allocate along the t…

mykaul opened the pull request:
…ask)

Use a single allocation for both the task and its stack, for locality.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
syncop: use flexible array for synctask's stack (allocate along the task)

Use a single allocation for both the task and its stack, for locality.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/georep-basic-rsync-ec.t

0 test(s) generated core 


3 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/00-geo-rep/georep-basic-dr-rsync-arbiter.t
./tests/00-geo-rep/georep-basic-rsync-ec.t
https://build.gluster.org/job/gh_centos7-regression/3000/

pranithk commented on libglusterfs/src/syncop.c:
```diff
@@ -474,7 +469,16 @@ synctask_create(struct syncenv *env, size_t stacksize, synctask_fn_t fn,
+    if (stacksize <= 0) {
+        newtask = GF_CALLOC(1, sizeof(struct synctask) + env->stacksize,
+                            gf_common_mt_syncstack);
+        newtask->ctx.uc_stack.ss_size = env->stacksize;
```
This should be gf_common_mt_synctask

mykaul replied:
Thanks, fixed.

pranithk commented on libglusterfs/src/syncop.c:
```diff
@@ -474,7 +469,16 @@ synctask_create(struct syncenv *env, size_t stacksize, synctask_fn_t fn,
+        newtask->ctx.uc_stack.ss_size = env->stacksize;
+    } else {
+        newtask = GF_CALLOC(1, sizeof(struct synctask) + stacksize,
+                            gf_common_mt_syncstack);
```
same here

mykaul replied:
Thanks, fixed.

xhernandez commented on libglusterfs/src/syncop.c:
```diff
@@ -474,7 +469,16 @@ synctask_create(struct syncenv *env, size_t stacksize, synctask_fn_t fn,
 
-    newtask = GF_CALLOC(1, sizeof(*newtask), gf_common_mt_synctask);
+    if (stacksize <= 0) {
+        newtask = GF_CALLOC(1, sizeof(struct synctask) + env->stacksize,
```
I would try to use `GF_MALLOC()`. The stack is of significant size and there's no reason to clear it. Without initializing it it's even possible that some portions of the stack won't be mapped until needed, reducing the physical memory utilization.

mykaul replied:
ACK - hoping I've initialized all members correctly!

xhernandez replied:
I think the initialization is fine. At least I haven't seen any issue.

A possible alternative for other cases (specially useful when most of the fields are initialized to 0) is this:

```c
ptr = malloc(sizeof(type) + extra);
memset(ptr, 0, sizeof(type));
```

Commit:
Fix review comments:
- correct memory type
- use MALLOC instead of CALLOC, which required properly initializing all structure's variables (which I hope I did correctly)

mykaul commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/replicate/do-not-reopen-fd.t

0 test(s) generated core 


2 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/bugs/replicate/do-not-reopen-fd.t
https://build.gluster.org/job/gh_centos7-regression/3003/

mykaul commented:
/run regresion

mykaul commented:
Unsure if the failures in regression tests are related to this PR.

mykaul commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/glusterd/check_elastic_server.t

0 test(s) generated core 


2 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/bugs/glusterd/check_elastic_server.t
https://build.gluster.org/job/gh_centos7-regression/3018/

mykaul commented on libglusterfs/src/syncop.c:
```diff
@@ -474,10 +469,22 @@ synctask_create(struct syncenv *env, size_t stacksize, synctask_fn_t fn,
+    if (stacksize <= 0) {
+        newtask = GF_MALLOC(sizeof(struct synctask) + env->stacksize,
+                            gf_common_mt_synctask);
+        newtask->ctx.uc_stack.ss_size = env->stacksize;
```
This, and line 479, are wrong - need to check (as we do in line 482) that newtask != NULL

Commit:
Fix review comments:
1. Ensure newtask is not NULL *before* accessing its variables!
2. memset synctask structure, to ensure everything is zero'ed out initially (as was before)
- this one I hope will get rid of some CI failures (still unsure they are related)

mykaul commented:
/run regression

mykaul commented:
@xhernandez I hope it's ok now
